presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java

```diff
@@ -151,102 +155,203 @@ private boolean advanceToNextRowGroup()
+        IntList offsets = new IntArrayList();
+        BooleanList valueIsNull = new BooleanArrayList();
+        calculateCollectionsOffsets(toArray(path), offsets, valueIsNull, keyDefinitionLevels, keyRepetitionLevels);
+        return ((MapType) type).createBlockFromKeyValue(valueIsNull.toBooleanArray(), offsets.toIntArray(), blocks[0], blocks[1]);
     }
 
     public Block readStruct(Type type, List<String> path)
             throws IOException
     {
-        return readStruct(type, path, new IntArrayList());
+        return readStruct(type, path, new IntArrayList(), new IntArrayList());
     }
 
-    private Block readStruct(Type type, List<String> path, IntList elementOffsets)
+    private Block readStruct(Type type, List<String> path, IntList definitionLevels, IntList repetitionLevels)
             throws IOException
     {
-        List<TypeSignatureParameter> parameters = type.getTypeSignature().getParameters();
-        Block[] blocks = new Block[parameters.size()];
-        for (int i = 0; i < parameters.size(); i++) {
-            NamedTypeSignature namedTypeSignature = parameters.get(i).getNamedTypeSignature();
+        List<TypeSignatureParameter> fields = type.getTypeSignature().getParameters();
+        Block[] blocks = new Block[fields.size()];
+        IntList fieldDefinitionLevels = null;
+        IntList fieldRepetitionLevels = null;
+        for (int i = 0; i < fields.size(); i++) {
+            fieldDefinitionLevels = new IntArrayList();
+            fieldRepetitionLevels = new IntArrayList();
+            NamedTypeSignature namedTypeSignature = fields.get(i).getNamedTypeSignature();
             Type fieldType = typeManager.getType(namedTypeSignature.getTypeSignature());
             String name = namedTypeSignature.getName();
-            blocks[i] = readBlock(name, fieldType, path, new IntArrayList());
+            blocks[i] = readBlock(name, fieldType, path, fieldDefinitionLevels, fieldRepetitionLevels);
+        }
+        BooleanList structIsNull = new BooleanArrayList();
+        IntList structOffsets = new IntArrayList();
+        calculateStructOffsets(toArray(path), structOffsets, structIsNull, definitionLevels, repetitionLevels, fieldDefinitionLevels, fieldRepetitionLevels);
+        return new RowBlock(0, structIsNull.size(), structIsNull.toBooleanArray(), structOffsets.toIntArray(), blocks);
+    }
+
+    /**
+     * Each struct has three variants of presence:
+     * 1) Struct is not defined, because one of it's optional parent fields is null
+     * 2) Struct is null
+     * 3) Struct is defined and not empty. In this case offset value is increased by one.
+     * <p>
+     * One of the struct's field repetition/definition levels are used to calculate offsets of the struct.
+     */
+    private void calculateStructOffsets(String[] path, IntList structOffsets, BooleanList structIsNull, IntList definitionLevels, IntList repetitionLevels, IntList fieldDefinitionLevels, IntList fieldRepetitionLevels)
```

Comment:
`definitionLevels` and `repetitionLevels` are for the struct I guess, so let's rename them as such. e.g., `structDefinitionLevels` ...